AI/src/search/mcts/nodes/PNMCTSNode.java

```diff
@@ -0,0 +1,339 @@
+        }
+    	
+    	evaluate();
+        setProofAndDisproofNumbers();
+        
+        childrenPNSSelectionTerms = new double[numLegalMoves()];
+    }
+    
+    //-------------------------------------------------------------------------
+    
+    /**
+     * Evaluates a node as in PNS according to L. V. Allis' 
+     * "Searching for Solutions in Games and Artificial Intelligence"
+     */
+    public void evaluate() 
+    {
+        if (this.context.trial().over()) 
+        {
+            if (RankUtils.utilities(this.context)[rootPlayer] == 1.0) 
+            {
+                this.proofValue = PNMCTSNodeValues.TRUE;
+            } 
+            else 
+            {
+                this.proofValue = PNMCTSNodeValues.FALSE;
+            }
+        } 
+        else 
+        {
+            this.proofValue = PNMCTSNodeValues.UNKNOWN;
+        }
+    }
+    
+    /**
+     * Sets the proof and disproof values of the current node as it is done for 
+     * PNS in L. V. Allis' "Searching for Solutions in Games and Artificial 
+     * Intelligence". Set differently depending on if the node has children yet.
+     *
+     * @return Returns true if something was changed and false if not. 
+     * Used to improve PN-MCTS speed.
+     */
+    public boolean setProofAndDisproofNumbers() 
+    {
+        if (legalMoves.length > 0) 
+        {
+        	// Not a terminal node
+        	double proof;
+        	double disproof;
+        	switch (type)
+        	{
+			case AND_NODE:
+				proof = 0.0;
+				disproof = Double.POSITIVE_INFINITY;
+				for (final BaseNode child : children)
+				{
+					final PNMCTSNode childNode = (PNMCTSNode) child;
+					if (childNode != null)
+					{
+						proof += childNode.proofNumber;
+						
+						if (childNode.disproofNumber < disproof)
+						{
+							disproof = childNode.disproofNumber;
+						}
+					}
+					else
+					{
+						// An unexpanded child
+						// TODO verify that this is indeed what we want to do?
+						// NOTE: addition for proof, setting for disproof
+						proof += 1.0;
+						disproof = 1.0;
+					}
+				}
+
+                // If nothing changed return false
+                if (this.proofNumber == proof && this.disproofNumber == disproof) 
+                {
+                    return false;
+                }
+                
+				this.proofNumber = proof;
+				this.disproofNumber = disproof;
+				return true;
+			case OR_NODE:
+				disproof = 0.0;
+				proof = Double.POSITIVE_INFINITY;
+				
+				for (final BaseNode child : children)
+				{
+					final PNMCTSNode childNode = (PNMCTSNode) child;
+					if (childNode != null)
+					{
+						disproof += childNode.disproofNumber;
+						
+						if (childNode.proofNumber < proof)
+						{
+							proof = childNode.proofNumber;
+						}
+					}
+					else
+					{
+						// An unexpanded child
+						// TODO verify that this is indeed what we want to do?
+						// NOTE: addition for disproof, setting for proof
```

Comment:
Jakub please check if this is what we want for children that have not been added to tree yet